poetry/console/commands/plugin/add.py

```diff
@@ -0,0 +1,263 @@
+        from poetry.puzzle.provider import Provider
+        from poetry.repositories.installed_repository import InstalledRepository
+        from poetry.repositories.repository import Repository
+        from poetry.utils.env import EnvManager
+
+        plugins = self.argument("plugins")
+
+        # Plugins should be installed in the system env to be globally available
+        system_env = EnvManager.get_system_env()
+
+        env_dir = Path(
+            os.getenv("POETRY_HOME") if os.getenv("POETRY_HOME") else system_env.path
+        )
+
+        # We check for the plugins existence first.
+        if env_dir.joinpath("pyproject.toml").exists():
+            pyproject = tomlkit.loads(
+                env_dir.joinpath("pyproject.toml").read_text(encoding="utf-8")
+            )
+            poetry_content = pyproject["tool"]["poetry"]
+            existing_packages = self.get_existing_packages_from_input(
+                plugins, poetry_content, "dependencies"
+            )
+
+            if existing_packages:
+                self.notify_about_existing_packages(existing_packages)
+
+            plugins = [plugin for plugin in plugins if plugin not in existing_packages]
+
+        if not plugins:
+            return 0
+
+        plugins = self._determine_requirements(plugins)
+
+        # We retrieve the packages installed in the system environment.
+        # We assume that this environment will be a self contained virtual environment
+        # built by the official installer or by pipx.
+        # If not, it might lead to side effects since other installed packages
+        # might not be required by Poetry but still taken into account when resolving dependencies.
+        installed_repository = InstalledRepository.load(
+            system_env, with_dependencies=True
+        )
+        repository = Repository()
+
+        root_package = None
+        for package in installed_repository.packages:
+            if package.name in Provider.UNSAFE_PACKAGES:
+                continue
+
+            if package.name == "poetry":
+                root_package = ProjectPackage(package.name, package.version)
+                for dependency in package.requires:
+                    root_package.add_dependency(dependency)
+
+                continue
+
+            repository.add_package(package)
+
+        root_package.python_versions = ".".join(
+            str(v) for v in system_env.version_info[:3]
+        )
+        # We create a `pyproject.toml` file based on all the information
+        # we have about the current environment.
+        self.create_pyproject_from_package(root_package, env_dir)
+
+        # We add the plugins to the dependencies section of the previously
+        # created `pyproject.toml` file
+        pyproject = tomlkit.loads(
+            env_dir.joinpath("pyproject.toml").read_text(encoding="utf8")
+        )
+        poetry_dependency_section = pyproject["tool"]["poetry"]["dependencies"]
```

Comment:
We should really use the pyproject classes here and avoid reimplementing certain things. Would be great to see that be the pattern for cases where we interact with the pyproject file.

Reply:
Changed :-)